### ISSUE TITLE
Properly process FCS files taken with high-resolution ADCs

### ISSUE THREAD

**Issue by castillohair** [taborlab/FlowCal#192]
We need to make sure that these files can pass through the Excel UI workflow, including MEF, at the very least.

These files are important because new instruments do not have logarithmic amplifiers. Instead, they pass PMT signals through a very high-resolution (18-bit typically) ADC, and store the event signals as floating points. Not supporting these files would reduce our audience significantly.

**Comment by castillohair** [comment]
One big issue here is the `domain` and `hist_bin_edges` properties of `FCSData`. In floating point, the range keyword `$PnR` can be really big. For our new test files, `PnR = 262144`. This can lead to slow performance and memory problems, particularly when plotting 2d density plots or performing density gating. Also, we have seen that some events can be less than zero. And because floating points have (conceptually at least) infinite resolution, `domain` kind of loses its meaning.

One solution is to eliminate `domain` and `hist_bin_edges` when data is in floating point, and give reasonable defaults for number of bins to plotting and density gating functions. I'm open to alternative suggestions.

**Comment by JS3xton** [comment]
Where are `domain` and `hist_bin_edges` used?

Perhaps `hist_bin_edges` can be automatically subsampled in some
intelligent way (that was its original purpose after all).

Can `domain` be a generator? Otherwise it feels like we only use the
cardinality of the domain set; perhaps we just expose $PnR? Can't remember
all the places where `domain` is used.

**Comment by castillohair** [comment]
`domain` is used:
- `high_low` gate, to get the default low/high values.
- mef module, when choosing which peaks to discard based on which events are saturated.
- plot module, to get the `xrange` and `yrange` of several plots.

In all of these cases, only the minimum and the maximum values are used.

`hist_bin_edges` is used:
- density gate
- `hist` and `density2d` in the `plot` module
- `excel_ui`, to produce the histograms tab

**Comment by castillohair** [comment]
I was thinking about discarding both `domain` and `hist_bin_edges`; and incorporating new arguments `range` and `resolution`. `range` would be a 2-element list that would originally go from 0 to `$PnR - 1`, and would be modified with each transformation in a similar way to what `domain` and `hist_bin_edges` do now. `resolution` would be copied from `$PnR` and never modified.

`domain` and `hist_bin_edges` were meant to solve the problem of non-uniform bin spacing after transformation. However, I think we now have a better solution. The Excel UI currently checks `amplification_type` to decide whether to scale some axes linearly or logarithmically, using the `log` argument of either `plot.hist1d` or `plot.density2d`. I think the `log` argument should also indicate the plotting functions to automatically generate bins spaced either linearly or logarithmically, using `range` as the lower and higher limits.

**Comment by castillohair** [comment]
About plotting functions, I also think the `div` argument could be dropped, and `bins` can be an integer directly indicating the number of bins to be generated from `range`. If `bins` is None, the function can use `FCSData.resolution` as the number of bins. `bins` can have a number, such as 256 or 1024, as a default, which would make things easier when dealing with super high resolutions such as when using floating points. The same idea goes to `gate.density2d`.

**Comment by castillohair** [comment]
Also, `FCSData` could still have `domain` and `hist_bin_edges` as functions that would generate bins from `range` with some specified resolution. This would then be used by plotting functions and density gating.

**Comment by castillohair** [comment]
We agreed about the following:
- Eliminate `div` from plotting functions.
- Use `endpoints` in place of what I proposed as `range`
- Use `resolution` as I proposed.
- @JS3xton is in favor of keeping `hist_bin_edges` as it is right now, but limit its resolution when `FCSData` is created. I think `hist_bin_edges` should be a function that generates bins from `endpoints` and `resolution`. An optional `nbins` argument to this function will specify the number of bins to generate, in which case the function ignores `resolution`. I'll try to make this function, but if it doesn't work well I'll go with John's idea.
- This issue will be narrowed down to deal with high-resolution ADCs, and another will be open to deal with linear MEFing.
